cli/internal/turbopath/turbopath.go

```diff
@@ -0,0 +1,180 @@
+type RelativeUnixPathInterface interface {
+	filePathStamp()
+	relativePathStamp()
+	unixPathStamp()
+
+	ToString() string
+}
+
+// RepoRelativeUnixPathInterface specifies the members that mark an interface
+// for structurally typing so the Go compiler can understand it.
+type RepoRelativeUnixPathInterface interface {
+	filePathStamp()
+	relativePathStamp()
+	repoRelativeStamp()
+	unixPathStamp()
+
+	ToString() string
+}
+
+// AbsoluteSystemPathInterface specifies the members that mark an interface
+// for structurally typing so the Go compiler can understand it.
+type AbsoluteSystemPathInterface interface {
+	filePathStamp()
+	absolutePathStamp()
+	systemPathStamp()
+
+	ToString() string
+}
+
+// RelativeSystemPathInterface specifies the members that mark an interface
+// for structurally typing so the Go compiler can understand it.
+type RelativeSystemPathInterface interface {
+	filePathStamp()
+	relativePathStamp()
+	systemPathStamp()
+
+	ToString() string
+}
+
+// RepoRelativeSystemPathInterface specifies the members that mark an interface
+// for structurally typing so the Go compiler can understand it.
+type RepoRelativeSystemPathInterface interface {
+	filePathStamp()
+	relativePathStamp()
+	repoRelativeStamp()
+	systemPathStamp()
+
+	ToString() string
+}
+
+// AbsolutePathInterface specifies additional dimensions that a particular
+// object can have that are independent from each other.
+type AbsolutePathInterface interface {
+	filePathStamp()
+	absolutePathStamp()
+
+	ToString() string
+}
+
+// RelativePathInterface specifies additional dimensions that a particular
+// object can have that are independent from each other.
+type RelativePathInterface interface {
+	filePathStamp()
+	relativePathStamp()
+
+	ToString() string
+}
+
+// RepoRelativePathInterface specifies additional dimensions that a particular
+// object can have that are independent from each other.
+type RepoRelativePathInterface interface {
+	filePathStamp()
+	relativePathStamp()
+	repoRelativeStamp()
+
+	ToString() string
+}
+
+// UnixPathInterface specifies additional dimensions that a particular
+// object can have that are independent from each other.
+type UnixPathInterface interface {
+	filePathStamp()
+	unixPathStamp()
+
+	Rel(UnixPathInterface) (RelativeUnixPath, error)
+	ToSystemPath() SystemPathInterface
+	ToUnixPath() UnixPathInterface
+	ToString() string
+}
+
+// SystemPathInterface specifies additional dimensions that a particular
+// object can have that are independent from each other.
+type SystemPathInterface interface {
+	filePathStamp()
+	systemPathStamp()
+
+	Rel(SystemPathInterface) (RelativeSystemPath, error)
+	ToSystemPath() SystemPathInterface
+	ToUnixPath() UnixPathInterface
+	ToString() string
+}
+
+// FilePathInterface specifies additional dimensions that a particular
+// object can have that are independent from each other.
+type FilePathInterface interface {
+	filePathStamp()
+
+	ToSystemPath() SystemPathInterface
```

Comment:
Earlier iterations of this PR *extensively* used `FilePathInterface` but I eventually decided to push the types up to call sites where I already knew what they were.

This is one of the takeaways for me: the more-general interfaces will primarily be used in situations where you're integrating with an existing repository. That forces you to make sure that everything that you do is safe while making it possible to avoid pushing all of the types up through the call sites.